src/Data/TypeRep/CacheMap.hs

```diff
@@ -0,0 +1,147 @@
+empty :: TypeRepMap f
+empty = TypeRepMap mempty mempty mempty
+
+-- | Inserts the value with its type as a key.
+insert :: forall a f . Typeable a => a -> TypeRepMap f -> TypeRepMap f
+insert = undefined
+
+-- | Looks up the value at the type.
+-- >>> let x = lookup $ insert (11 :: Int) empty
+-- >>> x :: Maybe Int
+-- Just 11
+-- >>> x :: Maybe ()
+-- Nothing
+lookup :: forall a f . Typeable a => TypeRepMap f -> Maybe (f a)
+lookup tVect =  fromAny . (anys tVect V.!)
+            <$> cachedBinarySearch (typeRepFingerprint $ typeRep $ Proxy @a)
+                             (fingerprintAs tVect)
+                             (fingerprintBs tVect)
+
+-- | Returns the size of the 'TypeRepMap'.
+size :: TypeRepMap f -> Int
+size = Unboxed.length . fingerprintAs
+
+-- | Binary searched based on this article
+-- http://bannalia.blogspot.com/2015/06/cache-friendly-binary-search.html
+-- with modification for our two-vector search case.
+cachedBinarySearch :: Fingerprint -> Unboxed.Vector Word64 -> Unboxed.Vector Word64 -> Maybe Int
+cachedBinarySearch (Fingerprint (W64# a) (W64# b)) fpAs fpBs = inline (go 0#)
+  where
+    go :: Int# -> Maybe Int
+    go i = case i <# len of
+        0# -> Nothing
+        _  -> let !(W64# valA) = Unboxed.unsafeIndex fpAs (I# i) in case a `ltWord#` valA of
+            0#  -> case a `eqWord#` valA of
+                0# -> go (2# *# i +# 2#)
+                _ -> let !(W64# valB) = Unboxed.unsafeIndex fpBs (I# i) in case b `ltWord#` valB of
+                    0# -> case b `eqWord#` valB of
+                        0# -> go (2# *# i +# 2#)
+                        _  -> Just (I# i)
+                    _ -> go (2# *# i +# 1#)
+            _ -> go (2# *# i +# 1#)
+
+    len :: Int#
+    len = let !(I# l) = Unboxed.length fpAs in l
+
+----------------------------------------------------------------------------
+-- Functions for testing and benchmarking
+----------------------------------------------------------------------------
+
+data TF f where
+  TF :: Typeable a => f a -> TF f
+
+fromF :: Typeable a => f a -> Proxy a
+fromF _ = Proxy
+
+fromList :: forall f . [TF f] -> TypeRepMap f
+fromList tfs = TypeRepMap (Unboxed.fromList fpAs) (Unboxed.fromList fpBs) (V.fromList ans)
+  where
+    (fpAs, fpBs) = unzip $ fmap (\(Fingerprint a b) -> (a, b)) fps
+    (fps, ans) = unzip $ breadthFirst $ fromListToTree $ sortWith fst $ map (fp &&& an) tfs
+
+    fp :: TF f -> Fingerprint
+    fp (TF x) = typeRepFingerprint $ typeRep $ fromF x
+
+    an :: TF f -> Any
+    an (TF x) = unsafeCoerce x
+
+----------------------------------------------------------------------------
+-- Tree
+----------------------------------------------------------------------------
+
+data Tree a = Leaf | Node a (Tree a) (Tree a)
+    deriving (Show)
+
+fromListToTree :: [a] -> Tree a
+fromListToTree [] = Leaf
+fromListToTree xs =
+    let len = length xs
+    in case splitAt (len `div` 2) xs of
+           ([], [])     -> Leaf
```

Comment:
Strictly speaking, this case can never happen. But we can't prove this to compiler. So it's okay to check this.